### ISSUE TITLE
Calling .NET Core WebAPI with Windows Authentication

### ISSUE THREAD

**Issue by Marsland** [elsa-workflows/elsa-core#2472]
I have an Elsa workflow using Designer that I would like to set up to call one of our internal APIs using Windows Authentication.
Elsa Server is running as an IIS application and has a domain level account set up for the AppPool, so the Server itself should be seen as a valid network account to the receiving API. The API we are calling is on the same server and implements CORS as well. I have tried using the SendHttpRequest action without an Authorization Header but receive a 401 response.

How do I set up the Authorization Header on the SendHttpRequest action to mimic what we do in a regular .NET API call to another API:
  HttpClient client = new(new HttpClientHandler() { UseDefaultCredentials = true });
            client.DefaultRequestHeaders.Accept.Clear();
            client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));

As always, thanks in advance for your splendid support and responses!

David Marsland

**Comment by mohdali** [comment]
Hi,

My colleague is working in a similar issue. We have not yet finalized it but the idea is that SendHttpRequest uses the th HttpClientFactory with a named client (name of the class). Then you can configure the factory as mentioned in MSDN docs https://docs.microsoft.com/en-us/aspnet/core/fundamentals/http-requests?view=aspnetcore-6.0#configure-the-httpmessagehandler

**Comment by Marsland** [comment]
Keep me posted if you will.

If I understand correctly, this would be a modification to the Elsa Server to allow using Header Propagation?

Thanks,
D

**Comment by mohdali** [comment]
No modification to the server. Should be manageable from your Startup class.

**Comment by Marsland** [comment]
That would be great. Let me know if you get all the details ironed out.

D

**Comment by mohdali** [comment]
@Marsland below snippet worked for us
```cs

services.AddHttpClient(nameof(SendHttpRequest)).ConfigurePrimaryHttpMessageHandler(_ => new HttpClientHandler
                {
                    Credentials = [put your CredentialsCache here]
                });
```

**Comment by Marsland** [comment]
I have added this snippet to my Startup with UseDefaultCredentials = true. I added a standard Send Http Request task to my workflow and configured it to point to a known good endpoint sending a POST with the appropriate payload defined.

I still get a 401 back. Seems like I'm missing something. I have the ELSA server running under a domain account on IIS so it should be getting seen as a domain account by the other API (which is inside our firewall).

Let me know if there's additional steps, or give me more details on how your setup is working (environment, etc.).

**Comment by mohdali** [comment]
In my case I used the CredentialsCache to directly provide the login credentials (username, password, domain) and it worked. I have not tested with Default Credentials

**Comment by Marsland** [comment]
At what location in the ConfigureServices section did you add this, before or after the AddElsa or AddElsaApiEndpoints? Still no luck here.

**Comment by mohdali** [comment]
It shouldn't matter, I  put at the end though.

Which version of Elsa are you using? We have tested this on 2.2.1 and 2.3.0

**Comment by Marsland** [comment]
We are running 2.3.0-preview.171 on our evaluation server.

I will try again later after taking care of some other issues.

**Comment by Marsland** [comment]
OK, I'm digging back into this and would like to ask a couple of questions the setup where you have this working.

Are you running the Server standalone in Kestrel or hosted as an application in IIS?

In either scenario, are you running the Elsa server with Windows Authentication enabled for itself?

I'm putting together a test bed Server on 2.3.0 with this mod running locally to better isolate where I'm not getting the desired result back outside of IIS, which is where we run our Elsa Server instance.

**Comment by Marsland** [comment]
Still no luck. Do I need to set Authorization or Request headers on my SendHttpRequest in the workflow?

**Comment by mohdali** [comment]
We are running inside IIS and yes windows authentication is enabled. However I don't think it matters because we are configruing the httpclient for sending requests.

Have you tried supplying the credentials using a Credentials Cache instead?
Check below example.
 https://beyondtechs.blogspot.com/2020/12/configure-httpclient-to-consume-windows.html

**Comment by Marsland** [comment]
Yes. After trying Default Credentials I tried setting a Credentials Cache (as per your previous note) with explicit user, password, domain with known good account information, but that did not work either.

Is this really the only thing you added:
services.AddHttpClient(nameof(SendHttpRequest)).ConfigurePrimaryHttpMessageHandler(_ => new HttpClientHandler
                {
                    Credentials = [put your CredentialsCache here]
                });
                
                
my code:
            Uri apis = new Uri("https://apis-dev.mycompany.com/");
            NetworkCredential nc = new NetworkCredential("someuser", "somepassword", "DOMAIN");
            CredentialCache cc = new CredentialCache();
            cc.Add(apis, "NTLM", nc);

            services.AddHttpClient(nameof(SendHttpRequest)).ConfigurePrimaryHttpMessageHandler(_ => new HttpClientHandler
            {
                //UseDefaultCredentials = true
                Credentials = cc
            });

The only other thing I see that I could toggle would be to add a specific api endpoint to the path but I wanted to just access any API on our DEV box from Elsa.

**Comment by mohdali** [comment]
This is my completed code snippet. We just created an options class to be able to set the credentials from appSettings.

This works well with Elsa 2.2.1.

Now I'm wondering if something broken in 2.3. I have to check again with latest preview to confirm.
```
var webCredentialsOptions = Configuration.GetSection(WebCredentialsOptions.WebCredentials)
                                                     .Get<WebCredentialsOptions>();
            if (webCredentialsOptions != null)
            {
                services.AddHttpClient(nameof(SendHttpRequest)).ConfigurePrimaryHttpMessageHandler(_ => new HttpClientHandler
                {
                    Credentials = webCredentialsOptions.GetCredentialCache()
                });
            }
```

**Comment by Marsland** [comment]
Hmmm. First off, I really appreciate your help and patience on this.

I can try creating a server instance with 2.2.1 as well.

Anything else that might be at variance? I assume you're using this for POSTs and GETs? Calling a RESTful endpoint with JSON? or something else?

Scratching my head but won't give up. Will take care of some other tasks and circle back to this and do let me know if you find out any more details.

D

**Comment by mohdali** [comment]
Hi,

I have tested with the latest preview (2.4-preview.406) and it works fine for me. I'm running using .Net Core 3.1 runtime.

Just wondering, is there any proxy server setup in your environment? http client uses the environment variables HTTP_PROXY. HTTPS_PROXY and NO_PROXY.

What is the error code are you getting when trying to call the API? is it 401 or something else? Could be useful to run it through Wireshark and check what is going on.

**Comment by Marsland** [comment]
We are on .NET Core 5, so that might be a factor. No proxy servers. All of this stuff is running in house inside our intranet.

Getting a straight 401 back. I've contemplated WireShark but this really has to boil down to something really simple and I often find it helpful to walk away from the problem periodically instead of wrestling with it. I think my next step is to capture the response and return it with the response from Elsa, but with a 401 it seems pretty obvious the credentials or the workaround aren't functioning as expected, although I've already swapped out the credentials once

I will try swapping the standalone project on my workstation to use Core 3.1 to see if that affects things, although I rather doubt it. It might be helpful if you could describe the API endpoint you're calling as I asked above, in our case we are calling to APIs hosted in IIS that we use with other applications, all RESTful WebAPI endpoints with standard Windows Authentication.

Again, appreciate your help and patience on this.

D

**Comment by Marsland** [comment]
One wrinkle might be we are using the CORS module for our APIs rather than use Cors in the Startup due to the problems with preflight requests.

**Comment by mohdali** [comment]
No worries. We are working on a similar scenario and will soon deploy it to production so I'm also interested to make sure we covered everything and not have any surprizes later.

In our case we are submitting POST request to a WebService hosted in IIS. That web service is configured to allow Windows Authentication only.

![image](https://user-images.githubusercontent.com/4145465/144430502-1fa305bc-cccc-4680-96d5-ec54ab98ab8c.png)

I'm sure the credentials cache is working because if I change it to something else it will not work and I end up with 401.

I'm not sure if CORS has any impact since CORS is enforced by the browser.

We did struggle sometime because we had a proxy in our environment and we ended up with 503 service unavailable for some time. However by checking in postman and Wireshark we could identify it was the proxy interfering.

I think it would save some time to try and troubleshoot in Wireshark and find what actually happens.
